src/appleseed/renderer/kernel/intersection/embreescene.cpp

```diff
@@ -0,0 +1,518 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "embreescene.h"
+
+// Standard headers.
+#include <memory>
+
+// Embree headers.
+#include <embree3/rtcore.h>
+
+// Foundation headers.
+#include "foundation/math/area.h"
+#include "foundation/math/fp.h"
+#include "foundation/math/intersection/raytrianglemt.h"
+#include "foundation/math/scalar.h"
+#include "foundation/math/transform.h"
+#include "foundation/utility/makevector.h"
+#include "foundation/utility/statistics.h"
+#include "foundation/utility/stopwatch.h"
+
+// appleseed.renderer headers.
+#include "renderer/kernel/intersection/intersectionsettings.h"
+#include "renderer/kernel/shading/shadingpoint.h"
+#include "renderer/kernel/shading/shadingray.h"
+#include "renderer/modeling/object/curveobject.h"
+#include "renderer/modeling/object/iregion.h"
+#include "renderer/modeling/object/meshobject.h"
+#include "renderer/modeling/object/object.h"
+#include "renderer/modeling/object/triangle.h"
+#include "renderer/modeling/scene/assembly.h"
+#include "renderer/modeling/scene/containers.h"
+
+using namespace foundation;
+using namespace std;
+
+
+namespace
+{
+
+using namespace renderer;
+
+struct GeometryData
+{
+    // Vertex data.
+    GVector3*               m_vertices;
+    size_t                  m_vertices_count;
+    size_t                  m_vertices_stride;
+
+    // Primitive data.
+    uint32*                 m_primitives;
+    size_t                  m_primitives_count;
+    size_t                  m_primitives_stride;
+
+    // Instance data.
+    size_t                  m_object_instance_idx;
+    uint32                  m_vis_flags;
+    size_t                  m_motion_steps_count;
+
+    RTCGeometryType         m_geometry_type;
+    RTCGeometry             m_geometry_handle;
+
+    ~GeometryData()
+    {
+        delete[] m_vertices;
+        delete[] m_primitives;
+        rtcReleaseGeometry(m_geometry_handle);
+    }
+};
+
+typedef map<unsigned int, shared_ptr<GeometryData>> GeometryDataContainer;
+typedef GeometryDataContainer::const_iterator       GeometryDataContainerIterator;
+
+inline void collect_triangle_data(
+    const ObjectInstance&   object_instance,
+    GeometryData&           geometry_data)
+{
+    assert(geometry_data.m_geometry_type == RTC_GEOMETRY_TYPE_TRIANGLE);
+
+    // Retrieve object space -> assembly space transform for the object instance.
+    const Transformd& transform = object_instance.get_transform();
+
+    // Retrieve the object.
+    Object& object = object_instance.get_object();
+
+    // TODO: remove regions.
+    // Retrieve the region kit of the object.
+    const Access<RegionKit> region_kit(&object.get_region_kit());
+
+    assert(region_kit->size() == 1);
+    const size_t region_idx = 0;
```

Comment:
For now we have regions so you'll need to iterate over them.

Reply:
MeshObject itself never produces regions...